src/lando/main/scm/git.py

```diff
@@ -0,0 +1,258 @@
+            f_msg.flush()
+            f_diff.write(diff)
+            f_diff.flush()
+
+            cmds = [
+                ["apply", f_diff.name],
+                ["add", "-A"],
+                [
+                    "commit",
+                    "--date",
+                    commit_date,
+                    "--author",
+                    commit_author,
+                    "--file",
+                    f_msg.name,
+                ],
+            ]
+
+            for c in cmds:
+                self._git_run(c, cwd=self.path)
+
+    @contextmanager
+    def for_pull(self) -> ContextManager:
+        """Context manager to prepare the repo with the correct environment variables set for pulling."""
+        yield self
+
+    @contextmanager
+    def for_push(self, requester_email: str) -> ContextManager:
+        """Context manager to prepare the repo with the correct environment variables set for pushing.
+
+        Args:
+            requester_email (str)
+        """
+        # We set the committer name to the requester's _email_ as this is the only piece
+        # of information about the user that we are comfortable making public. Names in
+        # the User objects are coming from LDAP, and may not be acceptable to use
+        # publicly.
+        os.environ[ENV_COMMITTER_NAME] = requester_email
+        os.environ[ENV_COMMITTER_EMAIL] = requester_email
+        logger.debug(
+            f"{ENV_COMMITTER_EMAIL} and {ENV_COMMITTER_NAME} set to {requester_email}"
+        )
+        try:
+            yield self
+        finally:
+            del os.environ[ENV_COMMITTER_NAME]
+            del os.environ[ENV_COMMITTER_EMAIL]
+
+    def head_ref(self) -> str:
+        """Get the current revision_id"""
+        return self._git_run("rev-parse", "HEAD", cwd=self.path)
+
+    def changeset_descriptions(self) -> list[str]:
+        """Retrieve the descriptions of commits in the repository.
+
+        Returns:
+            list[str]: A list of first lines of changeset descriptions.
+        """
+        command = ["log", "--format=%s", "@{u}.."]
+        return self._git_run(*command).splitlines()
+
+    def update_repo(self, pull_path: str, target_cset: Optional[str] = None) -> str:
+        """Update the repository to the specified changeset.
+
+        This method uses the Git commands to update the repository
+        located at the given pull path to the specified target changeset.
+
+        Args:
+            pull_path (str): The path to pull from.
+            target_cset (str): The target changeset to update the repository to.
+
+        Returns:
+            str: The target changeset
+        """
+        branch = target_cset or self.default_branch
+        self.clean_repo()
+        self._git_run("pull", "--prune", pull_path)
+        self._git_run("checkout", "--force", "-B", branch)
+        return self.head_ref()
+
+    def clean_repo(self, *, strip_non_public_commits: bool = True):
```

Comment:
nit: I don't think the "public" commit reference is fully applicable in git. Maybe we can choose a new name that applies to both repos but doesn't reference a legacy term. Maybe we should have separate `clean` and `reset` methods.

Reply:
Yep. I took it to mean “unpushed / local” commits. (The term seems to be “unpushed” for git https://duckduckgo.com/?t=ffab&q=git+unpublished+commits&ia=web)

Maybe `unpublished` commits? But then that's very similar to what we already have 😅